# version2.0

添加请求取消功能

```typescript
import axios from 'axios'
import type { AxiosError, CancelTokenSource, AxiosRequestConfig, AxiosResponse } from 'axios'
import { message } from 'antd'

const BASE_URL = process.env.NODE_ENV === 'production' ? '/data' : '/data'

interface RequestOption extends AxiosRequestConfig {
  headers?: AxiosRequestConfig['headers']
  data?: any
}

// 请求时的拦截
axios.interceptors.request.use(
  (config) => {
    if ([null, ''].includes(window.sessionStorage.getItem('TOKEN'))) {
      config.headers.token = window.sessionStorage.getItem('TOKEN')
    }
    // 发送请求之前做一些处理,loading...
    return config
  },
  async (error) => {
    // 当请求异常时做一些处理
    return await Promise.reject(error)
  }
)

// 响应时拦截
axios.interceptors.response.use(
  (response) => {
    return response
  },
  async (error) => {
    return await Promise.reject(error)
  }
)

async function errorState(error: AxiosError<any>): Promise<void> {
  // 判断http返回状态码错误

  // 取消请求不报错
  if (error?.code === 'ERR_CANCELED') return

  // 401 登录失效 需重新登录
  if (error.response?.status === 401) {
    await message.error(error.response?.data.msg ?? error.message)

    window.sessionStorage.setItem('TOKEN', '')
    window.location.pathname = '/login'
    return
  }

  // message.config({ maxCount: undefined }) // 最大提示框数量，超过限制被关闭
  console.log('errorState', error)
  void message.error(error.message)
}

function successState(response: any): void {
  const { data, status } = response
  // 统一判断后端返回的错误码
  if (status !== 200) {
    const error = new Error(data.msg)
    if (data.code === 2) {
      window.location.pathname = '/login'
      throw error
    } else throw error
  }
}

const request = async <T>(
  url: string,
  options: RequestOption,
  cancelTokenSource?: CancelTokenSource
): Promise<AxiosResponse<T>> => {
  const data = options.data ?? {}
  const params = options.method === 'GET' ? { params: data } : { data }

  delete options.data

  // http默认配置
  const httpDefaultOpts = {
    baseURL: BASE_URL,
    url,
    timeout: 30000,
    headers: options.headers ?? {},
    ...params,
    ...options
  }
  const token = window.sessionStorage.getItem('TOKEN')

  if (token !== null) {
    httpDefaultOpts.headers.token = token
  }

  if (cancelTokenSource == null) {
    cancelTokenSource = axios.CancelToken.source()
  }

  httpDefaultOpts.cancelToken = cancelTokenSource.token

  return await new Promise((resolve, reject) => {
    // console.log(httpDefaultOpts)
    axios(httpDefaultOpts)
      .then((resp) => {
        successState(resp)
        resolve(resp.data)
      })
      .catch((error) => {
        void errorState(error).then(() => {
          reject(error)
        })
      })
  })
}

export default request

```

使用 `request`

异常错误的处理

```typescript
export type AwaitWrap<T, K = any> = Promise<[T | null, AxiosError<K> | null]>

export const awaitwrap = async <T, K>(promise: Promise<AxiosResponse<T>>): AwaitWrap<T, K> => {
  return await promise
    .then(async (data: AxiosResponse<T>) => await ([data, null] as unknown as AwaitWrap<T, K>))
    .catch(async (err: AxiosError<K>) => await ([null, err] as unknown as AwaitWrap<T, K>))
}
```

`api`使用

```typescript
import type { CancelTokenSource } from 'axios'

export interface fetchApiParams {
  [name: string]: unknown
}

export const fetchApi = async (
  data: Partial<fetchApiParams>,
  cancelTokenSource?: CancelTokenSource
): AwaitWrap<API_REST<any>> => {
  return await awaitwrap(
    request(
      '/test/api',
      {
        method: 'GET',
        data
      },
      cancelTokenSource
    )
  )
}
```

页面中使用

```tsx
import axios, { type CancelTokenSource } from 'axios'

const cancelTokenSourceRef = useRef<CancelTokenSource | null>(null)

  useEffect(() => {
    console.log('fetchParams change', fetchParams)
    const loadData = async (): Promise<void> => {
    
      if (cancelTokenSourceRef.current != null) {
        cancelTokenSourceRef.current.cancel('Request cancelled by new request')
      }

      const newCancelTokenSource = axios.CancelToken.source()
      cancelTokenSourceRef.current = newCancelTokenSource
        
      void fetchTableData(fetchParams, cancelTokenSourceRef.current)
    }

    void loadData()

    return () => {
      cancelTokenSourceRef.current!.cancel('cancel')
    }
  }, [fetchParams])

 const fetchTableData = async (
    params: Partial<OrderPageParams> = {},
    cancelTokenSource?: CancelTokenSource
  ): Promise<void> => {
    setLoading(true)

    const _params = { ...params }
    const [res, err] = await fetchMethod(_params, cancelTokenSource)
    
    setLoading(false)
      
    if (err !== null) {
      console.log('order err', err)
    }
    
    if (res !== null) {
        console.log('res', res)
    // ...
    }
  }
```
